File: ABC188/E.cpp
# include <bits/stdc++.h>
using namespace std;

int main() {
    int N, M;
    cin >> N >> M;

    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }

    vector<vector<int>> road(N);
    for (int m = 0; m < M; m++){
        int x, y;
        cin >> x >> y;
        road[--x].push_back(--y);
    };

    const int INF = 1001001001;
    vector<int> dp(N, INF * 2);

    for (int i = 0; i < N; i++){
        for(auto j: road[i]){
            dp[j] = min(dp[j], A[i]);
            dp[j] = min(dp[j], dp[i]);
        }
    };

    int ans = - INF;

    for (int i = 0; i < N; i++) {
        ans = max(ans, A[i] - dp[i]);
    }
    cout << ans << endl;
}

File: ABC188/E.py
import sys

N, M = map(int, sys.stdin.readline().strip().split())

A = list(map(int, sys.stdin.readline().strip().split()))

road = []
for _ in range(M):
    X, Y = map(int, sys.stdin.readline().strip().split())
    road.append((X - 1, Y - 1))

road = sorted(road)

town = [[a, a] for a in A]

for x, y in road:
    if town[y][0] > town[x][0]:
        town[y][0] = town[x][0]

ans = 0

for a, b in town:
    if b - a > ans:
        ans = b - a

if ans != 0:
    print(ans)
    sys.exit()

ans = -(10 ** 10)
for x, y in road:
    ans = max(ans, A[y] - A[x])
print(ans)
